lib/Sema/TypeCheckPattern.cpp

```diff
@@ -816,6 +816,21 @@ bool TypeChecker::typeCheckPattern(Pattern *P, DeclContext *dc,
     TypedPattern *TP = cast<TypedPattern>(P);
     bool hadError = validateTypedPattern(*this, dc, TP, options, &resolver);
 
+    // Temporarily add a deprecation warning and when possible a fixit
+    // to replace ImplicitlyUnwrappedOptional<T> with T! so that we
+    // can eliminate the type ImplicitlyUnwrappedOptional in time.
+    auto *IUODecl = Context.getImplicitlyUnwrappedOptionalDecl();
```

Comment:
I think this should be moved to resolveType() in TypeCheckType.cpp. It's not just patterns where we want to prohibit explicitly spelling out IUO, but everywhere else too, eg

```
func foo(_: ImplicitlyUnwrappedOptional<Int>) {}
func foo() -> ImplicitlyUnwrappedOptional<Int> {}
func foo<T : Sequence>(_: T) where T.Element ==  ImplicitlyUnwrappedOptional<Int> {}
```

And so on.

This will also cover the extension case.

Reply:
Also

```
typealias T = ImplicitlyUnwrappedOptional<Int>
```

Actually right now even this is allowed, and it should not be

```
typealias T = Int!
```

Reply:
Yeah I had it implemented somewhere under there originally and then moved it to emit distinct diagnostics for the extension case, but now I realize doing so means we're emitting diagnostics for '!' as well.

Reply:
I don't think we want to ban ``typealias T = Int!`` if that's what you're suggesting. There's no great reason people shouldn't be able to do that and then declare 

```
func f(_ x: T) -> T { ... }
```

We do need to make sure these cannot appear anywhere that a directly written ``!`` appears.

Reply:
BTW, I view the fact that we can inject IUOs in places where they shouldn't be allowed as completely orthogonal and it's probably the last thing I'll actually get to unless it just happens to fall out.

Reply:
I see what you're saying, but I don't think there's much of benefit from defining such an alias. Also right now we can do stuff like

```
typealias T = Int!
func f(_: Array<T>) {}
```

Which should be banned, too.

Now we could allow the type alias but ban the array of it from being formed by looking through underlying types when we do our IUO checks, but that is more complex than just banning IUO in certain positions at the TypeRepr level.

I think the intent of the original IUO evolution proposal was that IUO is literally only allowed on parameters, returns, and properties, only directly and not via alias.

Reply:
I agree with Slava. The intent of SE-0054 was to make IUO not part of the type system at all.

Reply:
Piling on agreement with @slavapestov and @jrose-apple that SE-0054 implies that `!` should not be part of the type system at all. It's only allowed when directly attributable to a parameter, result type, or property.

Reply:
Right, a typealias is an alias at the Type level, not the TypeRepr level.